File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/자물쇠와 열쇠/Java/Solution.java
class Solution {
    public boolean solution(int[][] key, int[][] lock) {
        int N = lock.length;
        int M = key.length;
        for (int rowOffset = M; rowOffset > -N; rowOffset--) {
            for (int colOffset = M; colOffset > -N; colOffset--) {
                if (match(key, lock, rowOffset, colOffset)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean match(int[][] key, int[][] lock, int rowOffset, int colOffset) {
        int N = lock.length;
        int M = key.length;
        for (int rot = 0; rot < 4; rot++) {
            boolean matched = true;
            outer: for (int lockRow = 0; lockRow < N; lockRow++) {
                for (int lockCol = 0; lockCol < N; lockCol++) {
                    int keyPoint = 0;
                    int keyRow = lockRow + rowOffset;
                    int keyCol = lockCol + colOffset;

                    if (keyRow >= 0 && keyRow < M && keyCol >= 0 && keyCol < M) {
                        keyPoint = getKeyPoint(key, rot, keyRow, keyCol);
                    }

                    if ((lock[lockRow][lockCol] == 0 && keyPoint == 0) || (lock[lockRow][lockCol] == 1 && keyPoint == 1)) {
                        matched = false;
                        break outer;
                    }
                }
            }

            if (matched) {
                return true;
            }
        }

        return false;
    }

    private int getKeyPoint(int[][] key, int rot, int keyRow, int keyCol) {
        int M = key.length;
        return switch (rot) {
            case 0 -> key[keyRow][keyCol];
            case 1 -> key[M - keyCol - 1][keyRow];
            case 2 -> key[M - keyRow - 1][M - keyCol - 1];
            case 3 -> key[keyCol][M - keyRow - 1];
            default -> -1;
        };
    }
}

File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/자물쇠와 열쇠/solution.py
def solution(key, lock):
    # 전역 데이터
    M = len(key)
    N = len(lock)
    
    # 메서드
    def match(keyRowOffset, keyColOffset, rot):
        curKey = getCurKey(rot)
        for lockRow in range(N):
            keyRow = lockRow + keyRowOffset
            for lockCol in range(N):
                keyCol = lockCol + keyColOffset
                
                if not keyIsSafe(keyRow, keyCol):
                    if lock[lockRow][lockCol] == 0:
                        return False
                    continue
                
                if (lock[lockRow][lockCol] == 1 and curKey[keyRow][keyCol] == 1) or (lock[lockRow][lockCol] == 0 and curKey[keyRow][keyCol] != 1):
                    return False
        return True
    
    def getCurKey(rot):
        if rot == 0:
            return key
        
        curKey = [[0 for _ in range(M)] for _ in range(M)]
        for row in range(M):
            for col in range(M):
                if rot == 1:
                    curKey[col][M - row - 1] = key[row][col]
                elif rot == 2:
                    curKey[M - row - 1][M - col - 1] = key[row][col]
                else:
                    curKey[M - col - 1][row] = key[row][col]
                
        return curKey
    
    def keyIsSafe(row, col):
        return row >= 0 and row < M and col >= 0 and col < M
    
    # 메인 로직
    for keyRowOffset in range(M - 1, -N, -1):
        for keyColOffset in range(M - 1, -N, -1):
            for rot in range(4):
                if match(keyRowOffset, keyColOffset, rot):
                    return True
    
    return False
